### PR TITLE
[1.16.x] Added event for when a particle is spawned in the world

### PR DESCRIPTION
**Pull request by ChristianANoeddeboe**
Adds a new event, which is used when a particle is spawned in the world.

This allows players to temporarily replace or change a spawned vanilla particle.

An example scenario is increasing the damage the player does. So far a high damage hit on an entity will display many damage particles, and this will help modders reduce these to a reasonable amount.

Testmod has been added as ParticleWorldHandleEventTest.java.

### CODE REVIEW

File: patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch [hunk 1]
```diff
@@ -213,7 +213,35 @@
...
        explosion.func_77279_a(false);
        if (p_230546_12_ == Explosion.Mode.NONE) {
-@@ -1331,4 +1366,14 @@
+@@ -1019,12 +1054,14 @@
+    }
+ 
+    public <T extends IParticleData> int func_195598_a(T p_195598_1_, double p_195598_2_, double p_195598_4_, double p_195598_6_, int p_195598_8_, double p_195598_9_, double p_195598_11_, double p_195598_13_, double p_195598_15_) {
+-      SSpawnParticlePacket sspawnparticlepacket = new SSpawnParticlePacket(p_195598_1_, false, p_195598_2_, p_195598_4_, p_195598_6_, (float)p_195598_9_, (float)p_195598_11_, (float)p_195598_13_, (float)p_195598_15_, p_195598_8_);
++      net.minecraftforge.event.ServerParticleSpawnEvent event = net.minecraftforge.event.ForgeEventFactory.onWorldParticleSpawn(p_195598_1_, false, p_195598_2_, p_195598_4_, p_195598_6_, p_195598_8_, p_195598_9_, p_195598_11_, p_195598_13_, p_195598_15_);
++      if(event == null) return 0;
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch]: Formatting (missing space after `if`)

File: patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch [hunk 2]
```diff
@@ -213,7 +213,35 @@
...
+             ++i;
+          }
+       }
+@@ -1033,7 +1070,9 @@
+    }
+ 
+    public <T extends IParticleData> boolean func_195600_a(ServerPlayerEntity p_195600_1_, T p_195600_2_, boolean p_195600_3_, double p_195600_4_, double p_195600_6_, double p_195600_8_, int p_195600_10_, double p_195600_11_, double p_195600_13_, double p_195600_15_, double p_195600_17_) {
+-      IPacket<?> ipacket = new SSpawnParticlePacket(p_195600_2_, p_195600_3_, p_195600_4_, p_195600_6_, p_195600_8_, (float)p_195600_11_, (float)p_195600_13_, (float)p_195600_15_, (float)p_195600_17_, p_195600_10_);
++      net.minecraftforge.event.ServerParticleSpawnEvent event = net.minecraftforge.event.ForgeEventFactory.onWorldParticleSpawn(p_195600_2_, p_195600_3_, p_195600_4_, p_195600_6_, p_195600_8_, p_195600_10_, p_195600_11_, p_195600_13_, p_195600_15_, p_195600_17_);
++      if(event == null) return false;
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/world/server/ServerWorld.java.patch]: Formatting (missing space after `if`)

**Review by diesieben07 — CHANGES_REQUESTED** [review]

**Review by diesieben07 — APPROVED** [review]

**Comment by ichttt** [comment]
Is this event really needed? Is there no other way to cancel whatever spawen the particles? As this is an event that would fire quite often, we have to evalute if it is really worth the performance hit.

**Comment by ChristianANoeddeboe** [comment]
> 
> 
> Is this event really needed? Is there no other way to cancel whatever spawen the particles? As this is an event that would fire quite often, we have to evalute if it is really worth the performance hit.

For now it seems the only way to cancel or change vanilla particle spawns would be modifying the original code. In my use case a lot of damage particles are spawned when applying a high amount of damage, no matter how much health the mob has. This would allow me to cancel most of these, such if I only do 10% damage on a mob, it will not spam damage particles out.

**Comment by ichttt** [comment]
Sorry but I don't think this PR can be accepted in this way. This has a major performance penalty. We understand its a farily good place to hook in due to it being the central location. But it'd be better if the particles just wouldn't be spawned.

**Comment by ChristianANoeddeboe** [comment]
> 
> 
> Sorry but I don't think this PR can be accepted in this way. This has a major performance penalty. We understand its a farily good place to hook in due to it being the central location. But it'd be better if the particles just wouldn't be spawned.

Any alternative way you might see I can disable the particles spawning? I realize this will often be fired, but atm. I am not sure how I might affect particle spawning without modifying base classes. But I'm completely open to alternatives to this.

**Comment by sciwhiz12** [comment]
Hello! @ChristianANoeddeboe, please leave a comment whether you wish to continue this PR and update it to the latest active version of 1.18 (or leave an explanation why this PR's changes are not applicable to 1.18). If we do not hear back from you after some time, this PR will be closed for inactivity.
Additionally, if you do continue this PR, please rewrite your commits so the authors are properly in place for the CLA assistant to properly detect this PR's full compliance with the CLA.
If anyone else wishes to continue the work of this PR, please create a PR with a link to this PR for reference. 
Thank you for your contributions. 👋

**Comment by can-vaerks** [comment]
@sciwhiz12 
I don't mod minecraft anymore, and as it seems this didn't have a lot of interest, you can go ahead and close this one.

**Comment by ChristianANoeddeboe** [comment]
@sciwhiz12 
I don't mod minecraft anymore, and as it seems this didn't have a lot of interest, you can go ahead and close it.

**Pull request closed without merging**